drivers/pwm/pwm_imx.c

```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright (c) 2018, Diego Sueiro <diego.sueiro@gmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <pwm.h>
+#include <soc.h>
+#include <device_imx.h>
+
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_PWM_LEVEL
+#include <logging/sys_log.h>
+
+#define PWM_PWMSR_FIFOAV_4WORDS	0x4
+
+#define PWM_PWMCR_SWR(x) (((uint32_t)(((uint32_t)(x)) \
+				<<PWM_PWMCR_SWR_SHIFT))&PWM_PWMCR_SWR_MASK)
+
+#define DEV_CFG(dev) \
+	((const struct imx_pwm_config * const)(dev)->config->config_info)
+#define DEV_DATA(dev) \
+	((struct imx_pwm_data * const)(dev)->driver_data)
+#define DEV_BASE(dev) \
+	((PWM_Type *)(DEV_CFG(dev))->base)
+
+struct imx_pwm_config {
+	PWM_Type *base;
+	u16_t prescaler;
+};
+
+struct imx_pwm_data {
+	u32_t period_cycles;
+};
+
+
+static bool imx_pwm_is_enabled(PWM_Type *base)
+{
+	return PWM_PWMCR_REG(base) & PWM_PWMCR_EN_MASK;
+}
+
+static int imx_pwm_get_cycles_per_sec(struct device *dev, u32_t pwm,
+				       u64_t *cycles)
+{
+	PWM_Type *base = DEV_BASE(dev);
+	const struct imx_pwm_config *config = DEV_CFG(dev);
+
+	*cycles = get_pwm_clock_freq(base) >> config->prescaler;
+
+	return 0;
+}
+
+static int imx_pwm_pin_set(struct device *dev, u32_t pwm,
+			    u32_t period_cycles, u32_t pulse_cycles)
+{
+	PWM_Type *base = DEV_BASE(dev);
+	const struct imx_pwm_config *config = DEV_CFG(dev);
+	struct imx_pwm_data *data = DEV_DATA(dev);
+	unsigned int period_ms;
+	bool enabled = imx_pwm_is_enabled(base);
+	int wait_count = 0, fifoav;
+	u32_t cr, sr;
+
+
+	if ((period_cycles == 0) || (pulse_cycles > period_cycles)) {
+		SYS_LOG_ERR("Invalid combination: period_cycles=%d, "
+			    "pulse_cycles=%d", period_cycles, pulse_cycles);
+		return -EINVAL;
+	}
+
+	SYS_LOG_DBG("enabled=%d, pulse_cycles=%d, period_cycles=%d,"
+		    " duty_cycle=%d\n", enabled, pulse_cycles, period_cycles,
+		    (100 * pulse_cycles / period_cycles));
+
+	/*
+	 * i.MX PWMv2 has a 4-word sample FIFO.
+	 * In order to avoid FIFO overflow issue, we do software reset
+	 * to clear all sample FIFO if the controller is disabled or
+	 * wait for a full PWM cycle to get a relinquished FIFO slot
+	 * when the controller is enabled and the FIFO is fully loaded.
+	 */
+	if (enabled) {
+		sr = PWM_PWMSR_REG(base);
+		fifoav = PWM_PWMSR_FIFOAV(sr);
+		if (fifoav == PWM_PWMSR_FIFOAV_4WORDS) {
+			period_ms = (get_pwm_clock_freq(base) >>
+					config->prescaler) * MSEC_PER_SEC;
+			k_sleep(K_MSEC(period_ms));
+
+			sr = PWM_PWMSR_REG(base);
+			if (fifoav == PWM_PWMSR_FIFOAV(sr))
+				SYS_LOG_WRN("there is no free FIFO slot\n");
+		}
+	} else {
+		PWM_PWMCR_REG(base) = PWM_PWMCR_SWR(1);
+		do {
+			k_sleep(1);
```

Comment:
I didn't realize before that the k_busy_wait(500) is 500 microseconds, vs k_sleep(1) is one millisecond, but due to the default quantum is 10 milliseconds in Zephyr/ARM, it will sleep 10 milliseconds or more (unless kernel parameters are changed, you have maybe changed the system tick to 1 ms). Is this ok?